### ISSUE TITLE
[Question] `App` namespace redundant 

### ISSUE THREAD

**Issue by wtorsi** [symfony/flex#591]
Hi everybody, 

I've got one question about namespaces in app. 
My thought is that `App` namespace is redundant. Because:
- when you are developing _looks like huge_ applications it's not a good idea to store everything in one dir called `App`, it's better to separate logic in different dirs just for better understanding. For those purposes at this moment i removed 'App' namespace at all, and i feels lucky 😃

So, what's the reason to use `App` namespace? I think that app is looking better when all code is under one space called `root \ or /src`? 

Just have a look at my structure. Of course you need some tuning in `Kernel` & `composer`, but now, i think, it looks better. So what do you think? 

My idea is that _restricting_ (propagating) all code to `App` namespace it is a little bit hard to newer Symfony developers. It's better to just open new source (namespace) under root and be free.

![image](https://user-images.githubusercontent.com/2115840/71326498-e6c92a80-250c-11ea-9fca-087a72a1b4a7.png)

**Comment by Pierstoval** [comment]
Having the `App` namespace is important to make sure the `src/` directory is automatically discoverable with PSR-4 autoload rules. Having an autoload rule such as `"": "src/"` will have an impact on performances, that's why one needs to replace all namespaces if not using `App`.

Even if a concept of "src namespace" or something similar would be interesting to add to Flex, it would need a replacement of **all** recipes that are using this namespace, and manipulate **all** config files with `App\` in it in order to replace it with the proper namespace. Which is, even if feasible, a bit too cumbersome IMO.

Also, the MakerBundle depends on this `App` namespace, so it would need some work on this bundle too.

**Comment by alsar** [comment]
@Pierstoval what kind of performance penalty are we talking about when using `"": "src/"`?

**Comment by Pierstoval** [comment]
@alsar This kind: https://stackoverflow.com/questions/36158929/why-specify-the-namespace-when-using-psr-4-autoloading-with-composer

TL;DR: when you use `"": "src/"`, you tell Composer to search in `src/` for **all** classes you autoload. If you use `"App\\": "src/App"`, you tell composer to search there only for the `App` namespace, which is much faster. Especially considering the huge amount of classes that PHP loads during an HTTP request.

**Comment by alsar** [comment]
@Pierstoval Tnx for the explanation.  :)

**Comment by wtorsi** [comment]
Looking into `autoload.php`, composer splits classes by first letter of namespace in small chunks and then do search in those chunks by splitting class namespace. 

In case of `"":"src/"` our autoloaded class will be loaded just after usual search as fallback.  
As i can see for now it's only positive performance increase, for `src` classes there are no additional string operations.  
In case of APC those performance changes might only be visible on first load after APC flush.  
So i don't think that there is some performance impact, in case when you have only one fallback dir. 
But it's better to test it :) 


![image](https://user-images.githubusercontent.com/2115840/71360278-55a39380-25a0-11ea-943a-4bce2caaeefe.png)

**Comment by stof** [comment]
Having a namespace for your app code also avoids conflicts with class names defined either in PHP or in vendors (well, of course, nothing forbids a vendor from using `App` as top-level namespace, but then you would have a conflict only with that vendor rather than having to check for all namespace you use in your code, which are currently second-level ones)
